caffe2-cnn/mnist_nn_cpu.py

```diff
@@ -0,0 +1,197 @@
+  data = model.Cast(data_uint8, "data", to=core.DataType.FLOAT)
+  data = model.Scale(data, data, scale=float(1./256))
+  data = model.StopGradient(data, data)
+
+  # Create neural network structure
+  conv1 = model.Conv(data, 'conv1', dim_in=1, dim_out=assignments['conv1_dim'], kernel=5)
+  pool1 = model.MaxPool(conv1, 'pool1', kernel=2, stride=2)
+  conv2 = model.Conv(pool1, 'conv2', dim_in=assignments['conv1_dim'], dim_out=assignments['conv2_dim'], kernel=5)
+  pool2 = model.MaxPool(conv2, 'pool2', kernel=2, stride=2)
+  fc3 = model.FC(pool2, 'fc3', dim_in=assignments['conv2_dim'] * 4 * 4, dim_out=assignments['fc3'])
+  fc3 = model.Relu(fc3, fc3)
+  pred = model.FC(fc3, 'pred', assignments['fc3'], 10)
+  softmax = model.Softmax(pred, 'softmax')
+
+  # Add Training Operators
+  xent = model.LabelCrossEntropy([softmax, label], 'xent')
+  loss = model.AveragedLoss(xent, "loss")
+  accuracy = model.Accuracy([softmax, label], "accuracy")
+  model.AddGradientOperators([loss])
+  ITER = model.Iter("iter")
+  LR = model.LearningRate(
+    ITER, "LR", base_lr=-math.exp(assignments['log_learning_rate']), policy="step", stepsize=1, gamma=0.999 )
+  ONE = model.param_init_net.ConstantFill([], "ONE", shape=[1], value=1.0)
+  for param in model.params:
+    param_grad = model.param_to_grad[param]
+    model.WeightedSum([param, ONE, param_grad, LR], param)
+
+  return model
+
+# Evaluates the model and returns the accuracy
+def train_model(model):
+  workspace.RunNetOnce(model.param_init_net)
+  workspace.CreateNet(model.net)
+  total_iters = 100
+  for i in range(total_iters):
+    workspace.RunNet(model.net.Proto().name)
+    if(i % 5 == 0):
+      print("Iteration " + str(i) + " - Accuracy: " + str(workspace.FetchBlob('accuracy').item()))
+
+  return workspace.FetchBlob('accuracy').item()
+
+# Loops through retrieving suggestion, evaluating model, and reporting observation
+def sigopt_optimization_loop(conn, experiment):
+  for _ in range(experiment.observation_budget):
+    print('\n')
+    print("Optimization Loop #" + str(_))
+
+    # Fetch new suggestion
+    suggestion = conn.experiments(experiment.id).suggestions().create()
+    assignments = suggestion.assignments
+    print(assignments)
+
+    # Test suggestion
+    model = create_model(assignments)
+    accuracy = train_model(model)
+    print("Observed Accuracy: " + str(accuracy))
+
+    # Report observation
+    conn.experiments(experiment.id).observations().create(
+        suggestion=suggestion.id,
+        value=accuracy
+      )
+
+  # Get most accurate assignments
+  experiment = conn.experiments(experiment.id).fetch()
+  assignments = conn.experiments(experiment.id).best_assignments().fetch()
+
+  print('\n')
+  print("Best Assignments:")
+  print(assignments.data[0].assignments)
+
+  return assignments.data[0].assignments
+
+def main():
+  # Downloads data if it doesn't exist yet
+  data_setup()
+
+  conn = Connection(client_token='insert_token_here')
```

Comment:
this is buried pretty deep - might be best at the top of the file, and/or passable as a command-line arg / environment variable